OpenRA.Mods.Common/Activities/Attack.cs

```diff
@@ -114,16 +115,24 @@ public override bool Tick(Actor self)
 
 			useLastVisibleTarget = targetIsHiddenActor || !target.IsValidFor(self);
 
-			// If we are ticking again after previously sequencing a MoveWithRange then that move must have completed
-			// Either we are in range and can see the target, or we've lost track of it and should give up
-			if (wasMovingWithinRange && targetIsHiddenActor)
-				return true;
+			// If we are ticking again after previously sequencing a MoveWithRange then that move must have completed, and there is 3 situation:
+			// 1. we are in range and can see the target.
+			// 2. we've lost track of it and should give up.
+			// 3. we are stuck by immovable actors, we should give up.
+			// 4. we cannot reach the location to attack target, but we can retry.
+			if (wasMovingWithinRange)
+			{
+				if (targetIsHiddenActor)
+					return true;
+				else if (mobile != null && mobile.MoveResult == MoveResult.MovementStuck)
+					return true;
+				wasMovingWithinRange = false;
```

Comment:
For `Attack`, I am wondering if there's a meaningful difference whether the move completed and reached the destination or whether it got blocked. Maybe all we care is that it finished and so `if (lastMove?.MoveResult.IsComplete()) return true;` would suffice. We're either in range or we're not and that's that.

If the target moves whilst we are also moving, I guess we might finish our move successfully, but the target has moved along in the meantime and is now out of range? If so, we can move again.

But to prevent pathfinding spam, we can only retry if we haven't retired within say 120 ticks. If we move a long way to the target over 500 ticks and it shifts a bit, we'll try and catch up. But if the target is moving too fast for us, or alternatively if we're blocked in place and thus not making any progress, we can give up on the attack if we're trying to path again within 120 ticks.

So we retry forever if the attempts are >120 ticks apart. But we finish the activity if we make an attempt within 120 ticks of the previous attempt. We could also add a retry limit if we don't like the idea of it running forever.

We can adjust this policy. Overall we just need something that allows us to distinguish between "we're moving a lot and making progress" and "we're moving a lot and not making any progress" so that we know the energy we're spending on pathfinding is useful and not wasteful.